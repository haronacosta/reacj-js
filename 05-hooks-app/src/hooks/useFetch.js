import { useEffect, useRef, useState } from 'react';

const useFetch = (url) => {
  const isMounted = useRef(true);

  const [state, setState] = useState({
    data: null,
    loading: true,
    error: false,
  });

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    setState({ data: null, loading: true, error: false });

    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        if (isMounted.current) {
          setState({ loading: false, error: false, data });
        }
      });
  }, [url]);

  return state;
};

export default useFetch;
