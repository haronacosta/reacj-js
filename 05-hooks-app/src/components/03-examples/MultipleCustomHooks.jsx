import React from 'react';
import useFetch from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';

import '../02-useEffect/effect.css';

const MultipleCustomHooks = () => {
  const { counter, increment } = useCounter();

  const { loading, data } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );

  const { quote, author } = data ? data[0] : false;

  return (
    <div>
      <h1>Breaking Bad quotes</h1>

      <hr />

      {loading ? (
        <div className='alert alert-info text-center'>Loading...</div>
      ) : (
        <div>
          <figure className='text-end'>
            <blockquote className='blockquote'>
              <p>{quote}</p>
            </blockquote>
            <figcaption className='blockquote-footer'>{author}</figcaption>
          </figure>

          <button className='btn btn-primary' onClick={increment}>
            Next quote
          </button>
        </div>
      )}
    </div>
  );
};

export default MultipleCustomHooks;
