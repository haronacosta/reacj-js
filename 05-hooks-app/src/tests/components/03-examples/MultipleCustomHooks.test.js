import React from 'react';
import { shallow } from 'enzyme';
import MultipleCustomHooks from '../../../components/03-examples/MultipleCustomHooks';
import useFetch from '../../../hooks/useFetch';

jest.mock('../../../hooks/useFetch');

describe('Pruebas en <MultipleCustomHooks />', () => {
  test('Debe de mostrarse correctamente', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: false,
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper).toMatchSnapshot();
  });

  test('Debe de mostrar la informaciÃ³n', () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: 'Gabriel',
          quote: 'Hola Mundo',
        },
      ],
      loading: false,
      error: false,
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper.find('.alert').exists()).toBe(false);
    expect(wrapper.find('blockquote').text().trim()).toBe('Hola Mundo');
    expect(wrapper.find('.blockquote-footer').text().trim()).toBe('Gabriel');
  });
});
