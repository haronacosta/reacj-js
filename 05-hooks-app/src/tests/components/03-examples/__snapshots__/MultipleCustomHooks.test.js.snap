// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks /> Debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    Breaking Bad quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
</div>
`;
